cupy/cudnn.pyx

```diff
@@ -450,3 +494,335 @@ class DropoutStates(object):
+    assert dev_id == y.data.device.id
+
+    cdef float float_zero = 0, float_one = 1
+    cdef double double_zero = 0, double_one = 1
+    cdef size_t zero, one
+    if x.dtype == 'd':
+        zero = <size_t>&double_zero
+        one = <size_t>&double_one
+    else:
+        zero = <size_t>&float_zero
+        one = <size_t>&float_one
+
+    cdef bint use_tensor_core = _should_use_tensor_core(tensor_core, x.dtype)
+    cdef tuple conv_param = (pad, stride, x.dtype)
+
+    # cuDNN 7 supports dilation only in *_FWD_ALGO_IMPLICIT_GEMM, but
+    # it supports Tensor Cores only in *_FWD_ALGO_IMPLICIT_PRECOMP_GEMM.
+    if use_tensor_core:
+        for i in dilation:
+            if i > 1:
+                use_tensor_core = False
+                break
+
+    handle = get_handle()
+    x = core.ascontiguousarray(x)
+    W = core.ascontiguousarray(W)
+
+    # TODO(okuta) check performance
+    cdef size_t x_desc = cudnn.createTensorDescriptor()
+    cdef size_t y_desc = cudnn.createTensorDescriptor()
+    cdef size_t b_desc = cudnn.createTensorDescriptor()
+    cdef size_t filter_desc = cudnn.createFilterDescriptor()
+    cdef size_t conv_desc = cudnn.createConvolutionDescriptor()
+
+    cdef int algo
+    cdef size_t max_workspace_size = get_max_workspace_size()
+    cdef size_t workspace_size
+    try:
+        _create_tensor_nd_descriptor(x_desc, x, -1)
+        _create_tensor_nd_descriptor(y_desc, y, -1)
+        _create_filter_descriptor(filter_desc, W, cudnn.CUDNN_TENSOR_NCHW)
+        _create_convolution_descriptor(
+            conv_desc, pad, stride, dilation, groups, x.dtype,
+            cudnn.CUDNN_CROSS_CORRELATION, use_tensor_core)
+
+        workspace_size = max_workspace_size
+        if use_tensor_core:
+            # Only CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM
+            # supports Tensor-Core in cuDNN7.
+            algo = cudnn.CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM
+            workspace_size = cudnn.getConvolutionForwardWorkspaceSize(
+                handle, x_desc, filter_desc, conv_desc, y_desc, algo)
+            max_workspace_size = max(max_workspace_size, workspace_size)
+        elif auto_tune and _cudnn_version >= 5000:
+            algo, workspace_size = _get_algorithm_fwd(
+                x, W, y, conv_param, handle, x_desc, filter_desc,
+                conv_desc, y_desc, max_workspace_size)
+        else:
+            algo = cudnn.getConvolutionForwardAlgorithm(
+                handle, x_desc, filter_desc, conv_desc, y_desc,
+                cudnn.CUDNN_CONVOLUTION_FWD_SPECIFY_WORKSPACE_LIMIT,
+                max_workspace_size)
+        # TODO(okuta): allocate best size memory
+        workspace = memory.alloc(max_workspace_size)
+
+        cudnn.convolutionForward(
+            handle, one, x_desc, x.data.ptr, filter_desc, W.data.ptr,
+            conv_desc, algo, workspace.ptr, workspace_size, zero, y_desc,
+            y.data.ptr)
+
+        if b is None:
```

Comment:
I think it's better to write `if b is not None: <bias operations>`
It's not good to return earlier in ordinary execution path.